### PR TITLE
Clarify Redis Options and Update Defaults

### PR DESCRIPTION
**Pull request by kbrs-miaha**
There are 3 proposed changes:

* Clarify the settings required between *Redis*, *Redis Sentinel* and *Redis Cluster*
* Add examples explicitly for *Redis Sentinel*
* Change default for `enableCluster` to `false` to align with Bitnami Redis chart (`bitnami/redis`) [Potential Breaking Change]

The latter could be a breaking change, however the `bitnami/redis` chart used in the documentation does not run as a cluster, therefore this proposal aligns with the documentation (https://tyk.io/docs/tyk-oss/ce-helm-chart/).  However, the documentation also references a "Redis Cluster" address so there is an inconsistency.  The choice is to either change the documentation to reflect a *Redis Cluster* (e.g. `bitnami/redis-cluster`) or use the existing Bitnami Redis chart (`bitnami/redis`) but then update the defaults in this `values.yaml` file. This change proposes the latter.

## Related Issue
* Discussed and responding to Issue #217.

## Motivation and Context
* To use more consistent language regarding the use of *Redis* and it various configurations (e.g. *Redis*, *Redis Sentinel* and *Redis Cluster*.

## Test Coverage For This Change
* No formal testing - manually applied in an on-prem *Charmed* Kubernetes `v1.22.12` environment.

## Screenshots (if appropriate)
* N/A

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tyk-headless/values.yaml [hunk 1]
```diff
@@ -24,34 +24,47 @@ secrets:
...
+  # The addrs value will allow you to set your Redis addresses. 
+  #
+  # If you are using Redis (e.g. Bitnami Redis at bitnami/redis) then enter single
+  # endpoint.  If using "Redis Sentinel" update the port number (typically 26379).
+  #
+  # If using a "Redis Cluster" (e.g. bitnami/redis-cluster), you can list 
+  # the endpoints of the redis instances or use the cluster configuration endpoint.
+  # 
   # Default value: redis.{{ .Release.Namespace }}.svc.cluster.local:6379
+  #
```
Comment (buraksekili) [inline tyk-headless/values.yaml:36]: Please delete extra lines
```suggestion
```

File: tyk-headless/values.yaml [hunk 2]
```diff
@@ -24,34 +24,47 @@ secrets:
...
   # Enables SSL for Redis connection. Redis instance will have to support that.
   # Default value: false
   # useSSL: true
 
-  # The enableCluster value will allow you to indicate to Tyk whether you are
-  # running a Redis cluster or not.
-  # enableCluster: true
+  # If using "Redis Cluster" set enableCluster to true
+  # (e.g. if using bitnami/redis-cluster)
+  # enableCluster: false
```
Comment (buraksekili) [inline tyk-headless/values.yaml:56]: Please leave this part as it is for now.

File: tyk-headless/values.yaml [hunk 3]
```diff
@@ -24,34 +24,47 @@ secrets:
...
+  # If you are using Redis (e.g. Bitnami Redis at bitnami/redis) then enter single
+  # endpoint.  If using "Redis Sentinel" update the port number (typically 26379).
+  #
+  # If using a "Redis Cluster" (e.g. bitnami/redis-cluster), you can list 
+  # the endpoints of the redis instances or use the cluster configuration endpoint.
+  # 
   # Default value: redis.{{ .Release.Namespace }}.svc.cluster.local:6379
+  #
   # addrs:
-  #   - redis.tyk.svc.cluster.local:6379
```
Comment (buraksekili) [inline tyk-headless/values.yaml:32]: Can you please re-add this address? It is used for our simple charts https://tyk.io/docs/tyk-oss/ce-helm-chart/#evaluation-only-via-simple-redis-chart

File: tyk-headless/values.yaml [hunk 4]
```diff
@@ -24,34 +24,47 @@ secrets:
...
 redis:
-  # The addrs value will allow you to set your Redis addresses. If you are
-  # using a redis cluster, you can list the endpoints of the redis instances
-  # or use the cluster configuration endpoint.
+  # The addrs value will allow you to set your Redis addresses. 
+  #
+  # If you are using Redis (e.g. Bitnami Redis at bitnami/redis) then enter single
+  # endpoint.  If using "Redis Sentinel" update the port number (typically 26379).
+  #
+  # If using a "Redis Cluster" (e.g. bitnami/redis-cluster), you can list
```
Comment (buraksekili) [inline tyk-headless/values.yaml:32]:
```suggestion
  # If using a Redis Cluster (e.g. bitnami/redis-cluster), you can list 
```

File: tyk-headless/values.yaml [hunk 5]
```diff
@@ -24,34 +24,47 @@ secrets:
...
 
-  # The enableCluster value will allow you to indicate to Tyk whether you are
-  # running a Redis cluster or not.
-  # enableCluster: true
+  # If using "Redis Cluster" set enableCluster to true
+  # (e.g. if using bitnami/redis-cluster)
+  # enableCluster: false
 
-  # Enables sentinel connection mode for Redis. If enabled, sentinelPass and masterName values are required.
+  # If using "Redis Sentinel" set enableSentinel to true and provide both
```
Comment (buraksekili) [inline tyk-headless/values.yaml:58]: It'd be better to keep this comment generic as before, no need to be specific such as `"Redis Sentinel"`.

**Review by buraksekili — COMMENTED** [review]
Thanks @kbrs-miaha!! This looks great overall. I left my comments

Thank you for spotting this https://github.com/TykTechnologies/tyk-helm-chart/issues/217#issuecomment-1198003078, however; let's leave it as it is now. As you said, changing default values might be breaking change and it'd be better to handle in separate work.

File: tyk-headless/values.yaml (hunk 2 above)
Comment (kbrs-miaha) [reply tyk-headless/values.yaml:55]: Just to be clear, leave `enableCluster: false` for now?

Comment (kbrs-miaha) [reply tyk-headless/values.yaml:55]: I will return it back to true as per the original.

Comment (kbrs-miaha) [reply tyk-headless/values.yaml]: Please see revised commits.

File: tyk-headless/values.yaml (hunk 3 above)
Comment (kbrs-miaha) [reply tyk-headless/values.yaml:32]: Thank you for sharing the context; I have added this is a specific example.

File: tyk-headless/values.yaml (hunk 5 above)
Comment (kbrs-miaha) [reply tyk-headless/values.yaml]: Updated as per comments.

**Comment by kbrs-miaha** [comment]
@buraksekili - Thanks for the feedback - I will look to you for guidance on word choice and formatting so its consistent for you.  Happy to make any other edits required.

File: tyk-headless/values.yaml [hunk 6]
```diff
@@ -24,34 +24,48 @@ secrets:
   useSecretName: ""
 
 redis:
-  # The addrs value will allow you to set your Redis addresses. If you are
-  # using a redis cluster, you can list the endpoints of the redis instances
-  # or use the cluster configuration endpoint.
+  # The addrs value will allow you to set your Redis addresses. 
+  #
+  # If you are using Redis (e.g. Bitnami Redis at bitnami/redis) then enter single
+  # endpoint.  If using sentinel connection mode for Redis update the port number (typically 26379).
```
Comment (buraksekili) [inline tyk-headless/values.yaml:30]:
```suggestion
  # endpoint. If using sentinel connection mode for Redis, please update the port number (typically 26379).
```

**Review by buraksekili — COMMENTED** [review]
LGTM

**Comment by buraksekili** [comment]
> @buraksekili - Thanks for the feedback - I will look to you for guidance on word choice and formatting so its consistent for you. Happy to make any other edits required.

Thank you @kbrs-miaha for your effort 🙏  The PR looks good to me. Can you also please fix the linter issues? It complains about [`trailing spaces`](https://github.com/TykTechnologies/tyk-helm-chart/runs/7562597206?check_suite_focus=true).

@komalsukhani can you please review this PR as well?

**Review by buraksekili — APPROVED** [review]

**Comment by kbrs-miaha** [comment]
Great to hear - I believe I have eliminated the trailing spaces.

**Comment by kbrs-miaha** [comment]
As I'm doing on GitHub and not on `vim` it appears I didn't eliminate all of the trailing spaces.  Made a second attempt.

**Review by andrei-tyk — APPROVED** [review]

File: tyk-headless/values.yaml [hunk 7]
```diff
@@ -24,34 +24,48 @@ secrets:
...
+  # If using a Redis Cluster (e.g. bitnami/redis-cluster), you can list
+  # the endpoints of the redis instances or use the cluster configuration endpoint.
+  #
   # Default value: redis.{{ .Release.Namespace }}.svc.cluster.local:6379
   # addrs:
+  #   Example using tyk simple redis chart
   #   - redis.tyk.svc.cluster.local:6379
-  #   This is the the DNS name of the redis as set when using Bitnami
+  #   Example using bitnami/redis
+  #   - tyk-redis.tyk.svc.cluster.local:6379
```
Comment (komalsukhani) [inline tyk-headless/values.yaml:40]:
```suggestion
  #   - tyk-redis-master.tyk.svc.cluster.local:6379
```
For bitnami redis, we need to connect to `tyk-redis-master.tyk.svc.cluster.local:6379` address

File: tyk-headless/values.yaml [hunk 8]
```diff
@@ -24,34 +24,48 @@ secrets:
...
+  #   Example using bitnami/redis with sentinel
+  #   - tyk-redis.tyk.svc.cluster.local:26379
+  #   Example using bitnami/redis-cluster
   #   - tyk-redis-master.tyk.svc.cluster.local:6379
```
Comment (komalsukhani) [inline tyk-headless/values.yaml:44]:
```suggestion
  #   - tyk-redis-redis-cluster.tyk.svc.cluster.local:6379
```
For cluster mode, endpoint should be `tyk-redis-redis-cluster.tyk.svc.cluster.local:6379`

**Comment by kbrs-miaha** [comment]
I've accepted the suggestions regarding `redis-cluster` as I kept the original pattern, its the configuration I'm least familiar with but the changes seem logical.

**Comment by komalsukhani** [comment]
Thank you @kbrs-miaha for the PR!

One last thing and then we can merge the PR.
We have started using Changlelog file. Can you update your branch with latest master and add the one liner [here](https://github.com/TykTechnologies/tyk-helm-chart/blob/d09e0b5d499eadc04930ddd0606bfd62451f23dc/CHANGELOG.md?plain=1#L16)

**Comment by kbrs-miaha** [comment]
Apologies for the delay, I:

* Synced branch with master.
* Added 1 line into changelog.

**Comment by komalsukhani** [comment]
@kbrs-miaha Thank you for your contribution!!

**Comment by kbrs-miaha** [comment]
Thanks for supporting open source - I was happy to help and it's been an enlightening experience. Keep up the good work 🚀 !

**Pull request merged**